### PR TITLE
[FLINK-18258][hive] Implement SHOW PARTITIONS for Hive dialect

### PR DESCRIPTION
**Pull request by SteNicholas**
## What is the purpose of the change

*Currently Hive dialect supports `ADD PARTITION` and `DROP PARTITION`, lack of support for `SHOW PARTITIONS`. It's necessary to implement `SHOW PARTITIONS` for Hive dialect.*

## Brief change log

  - *Add `SqlShowPartitions` and `ShowPartitionsOperation` to support `SqlToOperationConverter` for `SHOW PARTITIONS` convert.*
  - *Method `executeOperation` of `TableEnvironmentImpl` support execution of `ShowPartitionsOperation`.*

## Verifying this change

  - *`FlinkHiveSqlParserImplTest` add method `testShowPartitions` to verify the parser of `SHOW PARTITIONS` dialect.*
  - *Method `testAddDropPartitions` of `HiveDialectITCase` add execution of `SHOW PARTITIONS` dialect to verify the result of the operation.*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

**Comment by SteNicholas** [comment]
> Thanks for contribution.
> 
> * Can you add case for sql-client?
> * Can you add case for full time partition `2020-02-02 05:06:07`.

OK, I would like to add the case you mentioned.

File: flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/connectors/hive/HiveDialectITCase.java [hunk 1]
```diff
@@ -450,6 +450,18 @@ public void testAddDropPartitions() throws Exception {
 		ObjectPath tablePath = new ObjectPath("default", "tbl");
 		assertEquals(2, hiveCatalog.listPartitions(tablePath).size());
 
+		List<Row> partitions = Lists.newArrayList(tableEnv.executeSql("show partitions tbl").collect());
```
[the remark this answers is not in the document]
Comment (SteNicholas) [reply flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/connectors/hive/HiveDialectITCase.java:453]: @lirui-apache OK, I would like to separate this case to another test case.

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/TableEnvironmentImpl.java [hunk 2]
```diff
@@ -1021,6 +1022,28 @@ private TableResult executeOperation(Operation operation) {
 			return buildShowResult("function name", listFunctions());
 		} else if (operation instanceof ShowViewsOperation) {
 			return buildShowResult("view name", listViews());
+		} else if (operation instanceof ShowPartitionsOperation) {
+			String exMsg = getDQLOpExecuteErrorMsg(operation.asSummaryString());
```
[the remark this answers is not in the document]
Comment (SteNicholas) [reply flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/TableEnvironmentImpl.java:1026]: @lirui-apache I thought that `SHOW PARTITIONS` is DQL. I will modify `getDQLOpExecuteErrorMsg` to `getDDLOpExecuteErrorMsg`.

File: flink-table/flink-sql-client/src/test/java/org/apache/flink/table/client/cli/SqlCommandParserTest.java [hunk 3]
```diff
@@ -299,6 +299,21 @@ public void testCommands() throws Exception {
 		}
 	}
 
+	@Test
+	public void testHiveCommands() throws Exception {
+		SqlParserHelper helper = new SqlParserHelper(SqlDialect.HIVE);
+		parser = helper.getSqlParser();
+		List<TestItem> testItems = Arrays.asList(
+			// show partitions
+			TestItem.invalidSql("SHOW PARTITIONS ", SqlExecutionException.class, "Encountered \"<EOF>\""),
```
[the remark this answers is not in the document]
Comment (SteNicholas) [reply flink-table/flink-sql-client/src/test/java/org/apache/flink/table/client/cli/SqlCommandParserTest.java:308]: @lirui-apache I would like to remove this case. Previously I use this test case to verfiy `SHOW PARTITIONS` without table name.

File: flink-table/flink-sql-client/src/test/java/org/apache/flink/table/client/cli/SqlCommandParserTest.java [hunk 4]
```diff
@@ -389,6 +409,15 @@ public static TestItem validSql(
 			return testItem;
 		}
 
+		public static TestItem validSql(
```
[the remark this answers is not in the document]
Comment (SteNicholas) [reply flink-table/flink-sql-client/src/test/java/org/apache/flink/table/client/cli/SqlCommandParserTest.java:412]: @lirui-apache Sorry for previously not finding the `TestItem::validSql` method that takes SQL dialect as a parameter. I would like to call this method.

File: flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/cli/CliClient.java [hunk 5]
```diff
@@ -549,6 +552,22 @@ private void callShowModules() {
 		terminal.flush();
 	}
 
+	private void callShowPartitions() {
+		final List<String> partitions;
+		try {
+			partitions = getShowResult("PARTITIONS");
```
[the remark this answers is not in the document]
Comment (SteNicholas) [reply flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/cli/CliClient.java]: @lirui-apache Yes, this need the table identifier and partition spec. I missed it.

File: flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/connectors/hive/HiveDialectITCase.java [hunk 6]
```diff
@@ -465,6 +465,49 @@ public void testAddDropPartitions() throws Exception {
...
+		tableEnv.executeSql("alter table tbl add partition (dt='2020-04-30',country='china') partition (dt='2020-04-30',country='us')");
+
+		ObjectPath tablePath = new ObjectPath("default", "tbl");
+		assertEquals(2, hiveCatalog.listPartitions(tablePath).size());
+
+		List<Row> partitions = Lists.newArrayList(tableEnv.executeSql("show partitions tbl").collect());
+		assertEquals(2, partitions.size());
+		assertTrue(partitions.toString().contains("dt=2020-04-30/country=china"));
+		assertTrue(partitions.toString().contains("dt=2020-04-30/country=us"));
+		partitions = Lists.newArrayList(tableEnv.executeSql("show partitions tbl partition (dt='2020-04-30')").collect());
```
Comment (lirui-apache) [inline flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/connectors/hive/HiveDialectITCase.java:480]: Also add a test case where the partition spec only contains `country`

File: flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/connectors/hive/HiveDialectITCase.java [hunk 7]
```diff
@@ -465,6 +465,49 @@ public void testAddDropPartitions() throws Exception {
...
+		assertTrue(partitions.toString().contains("dt=2020-04-30/country=china"));
+		assertTrue(partitions.toString().contains("dt=2020-04-30/country=us"));
+		partitions = Lists.newArrayList(tableEnv.executeSql("show partitions tbl partition (dt='2020-04-30',country='china')").collect());
+		assertEquals(1, partitions.size());
+		assertTrue(partitions.toString().contains("dt=2020-04-30/country=china"));
+
+		tableEnv.executeSql("alter table tbl drop partition (dt='2020-04-30',country='china'),partition (dt='2020-04-30',country='us')");
+		assertEquals(0, hiveCatalog.listPartitions(tablePath).size());
+
+		tableEnv.executeSql("alter table tbl add partition (dt='2020-04-30 01:02:03',country='china') partition (dt='2020-04-30 04:05:06',country='us')");
```
Comment (lirui-apache) [inline flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/connectors/hive/HiveDialectITCase.java:491]: Does a DATE column accept values like `'2020-04-30 01:02:03'`?

File: flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/connectors/hive/HiveDialectITCase.java [hunk 8]
```diff
@@ -465,6 +465,49 @@ public void testAddDropPartitions() throws Exception {
...
+		assertTrue(partitions.toString().contains("dt=2020-04-30 01:02:03/country=china"));
+		partitions = Lists.newArrayList(tableEnv.executeSql("show partitions tbl partition (dt='2020-04-30 04:05:06')").collect());
+		assertEquals(1, partitions.size());
+		assertTrue(partitions.toString().contains("dt=2020-04-30 04:05:06/country=us"));
+		partitions = Lists.newArrayList(tableEnv.executeSql("show partitions tbl partition (dt='2020-04-30 01:02:03',country='china')").collect());
+		assertEquals(1, partitions.size());
+		assertTrue(partitions.toString().contains("dt=2020-04-30 01:02:03/country=china"));
+
+		tableEnv.executeSql("alter table tbl drop partition (dt='2020-04-30 01:02:03',country='china'),partition (dt='2020-04-30 04:05:06',country='us')");
+		assertEquals(1, hiveCatalog.listPartitions(tablePath).size());
```
Comment (lirui-apache) [inline flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/connectors/hive/HiveDialectITCase.java:508]: I don't this is necessary. We're not testing add/drop partitions here.

File: flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/cli/CliClient.java [hunk 9]
```diff
@@ -532,6 +535,14 @@ private void callShowFunctions() {
 				.collect(Collectors.toList());
 	}
 
+	private List<String> getShowResult(String objectToShow, SqlCommandCall cmdCall) {
+		TableResult tableResult = executor.executeSql(sessionId, "SHOW " + objectToShow + " " + cmdCall.operands[0]);
```
Comment (lirui-apache) [inline flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/cli/CliClient.java:539]: I think `cmdCall.operands[0]` contains the whole SQL statement, no?

File: flink-table/flink-sql-client/src/test/java/org/apache/flink/table/client/cli/SqlCommandParserTest.java [hunk 10]
```diff
@@ -299,6 +300,20 @@ public void testCommands() throws Exception {
 		}
 	}
 
+	@Test
+	public void testHiveCommands() throws Exception {
+		SqlParserHelper helper = new SqlParserHelper(SqlDialect.HIVE);
+		parser = helper.getSqlParser();
```
Comment (lirui-apache) [inline flink-table/flink-sql-client/src/test/java/org/apache/flink/table/client/cli/SqlCommandParserTest.java:306]: Do we still need these?

File: flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/connectors/hive/HiveDialectITCase.java (hunk 6 above)
Comment (SteNicholas) [reply flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/connectors/hive/HiveDialectITCase.java:480]: @lirui-apache Yeah, I would like to add this case to verify.

File: flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/connectors/hive/HiveDialectITCase.java (hunk 8 above)
Comment (SteNicholas) [reply flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/connectors/hive/HiveDialectITCase.java:508]: @lirui-apache Okay, I will remove this unnecessary case.

File: flink-table/flink-sql-client/src/test/java/org/apache/flink/table/client/cli/SqlCommandParserTest.java (hunk 10 above)
Comment (SteNicholas) [reply flink-table/flink-sql-client/src/test/java/org/apache/flink/table/client/cli/SqlCommandParserTest.java:306]: @lirui-apache It's indeed unnecessary.

File: flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/connectors/hive/HiveDialectITCase.java (hunk 7 above)
Comment (SteNicholas) [reply flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/connectors/hive/HiveDialectITCase.java:491]: @lirui-apache I will modify the column to Timestamp type.

File: flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/cli/CliClient.java (hunk 9 above)
Comment (SteNicholas) [reply flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/cli/CliClient.java]: Yeah, I check the `parseBySqlParser` method, and confirm the `cmdCall.operands[0]` contains the whole SQL statement.

**Pull request merged**